plugin/mutate/source/dextool/plugin/mutate/backend/report/html/page_tree_map.d

```diff
@@ -1,60 +1,89 @@
 module dextool.plugin.mutate.backend.report.html.page_tree_map;
 
+import std.json : JSONValue;
 import arsd.dom : Document, Element, require, Table, RawSource;
 
 import dextool.plugin.mutate.backend.report.html : FileIndex;
 import dextool.plugin.mutate.backend.report.html.tmpl : tmplBasicPage;
 
+/* 
+* A JSON-like object that is more easily manipulated than JSONValue.
+* Has operator overloads for "index", "index assign" and "in" allowing the children list to be private
+* and cleaner construction of the object. 
+*/
+class JSONLike {
+    string name;
+    private JSONLike[string] children;
+    long locs = -1;
+    double score = -1;
+
+    JSONValue toJSONValue() {
+        JSONValue root = ["name" : name];
+
+        if (children.length) {
+            JSONValue[] descendants = [];
+            foreach (c; children) {
+                descendants ~= c.toJSONValue();
+            }
+            root.object["children"] = JSONValue(descendants);
+        }
+        if (locs != -1)
+            root.object["locs"] = JSONValue(locs);
+
+        if (score != -1)
+            root.object["score"] = JSONValue(score);
+
+        return root;
+    }
+
+    JSONLike opIndex(string i) pure nothrow {
+        return children[i];
+    }
+
+    JSONLike opIndexAssign(JSONLike value, string i) pure nothrow {
+        children[i] = value;
+        return value;
+    }
+
+    bool opBinaryRight(string op)(string a) {
+        if (op == "in") {
+            if (a in children)
+                return true;
+            return false;
+        }
+        return false;
+    }
+
+}
+
 @trusted auto makeTreeMapJSON(FileIndex[] files) {
-    import std.stdio : writeln;
     import std.array : array;
-    import std.range : enumerate;
-    import std.json : JSONValue;
     import std.path : pathSplitter;
-    import std.format;
 
-    JSONValue root = ["name" : "root"];
-    auto parent = root;
+    auto iRoot = new JSONLike();
+    iRoot.name = "root";
+
     foreach (f; files) {
         auto path = pathSplitter(f.display);
-
-        segments: foreach (i, seg; path.enumerate) {
-            if ("children" in parent) { // Parent has children -> look for segment in children
-                for (int j; j < parent["children"].array.length; j++) {
-                    auto child = parent["children"][j];
-                    
-                    if (child["name"].str == seg) {// Segment already added to JSON
-                        parent = child; // Update parent
-                        continue segments; // Go to next segment
-                    }
-                }
-            } 
-            else { // Parent has no children -> init list
-                parent.object["children"] = JSONValue[].init;
+        auto parent = iRoot;
+        foreach (seg; path) {
```

Comment:
The intermediary class turned out to be a lot easier to manipulate, resulting in a cleaner "makeTreeMapJSON" function.